src/interactive-window/interactiveWindow.ts

```diff
@@ -132,144 +119,137 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
+                this.owner,
+                this.notebookDocument,
+                undefined,
+                onStartKernel
+            );
+            this.currentKernelInfo.controller = kernel.controller;
+            this.currentKernelInfo.metadata = kernel.kernelConnectionMetadata;
+
+            const kernelEventHookForRestart = async (ev: 'willRestart' | 'willInterrupt') => {
+                if (ev === 'willRestart' && this.notebookDocument && this.currentKernelInfo.metadata) {
+                    this._insertSysInfoPromise = undefined;
+                    // If we're about to restart, insert a 'restarting' message as it happens
+                    void this.insertSysInfoMessage(this.currentKernelInfo.metadata, SysInfoReason.Restart);
                 }
-                this._kernelConnectionId = controller.id;
+            };
+            // Hook pre interrupt so we can stick in a message
+            kernel.addEventHook(kernelEventHookForRestart);
+            this.kernelDisposables.push(new Disposable(() => kernel.removeEventHook(kernelEventHookForRestart)));
+
+            // When restart finishes, rerun our initialization code
+            kernel.onRestarted(
+                async () => {
+                    traceInfoIfCI('Restart event handled in IW');
+                    this.fileInKernel = undefined;
+                    const cellPromise = Promise.resolve(
+                        this.notebookDocument.cellAt(this.notebookDocument.cellCount - 1)
+                    );
+                    try {
+                        await this.runInitialization(kernel, this.owner);
+                    } finally {
+                        this.finishSysInfoMessage(kernel, cellPromise, SysInfoReason.Restart);
+                    }
+                },
+                this,
+                this.kernelDisposables
+            );
+            this.kernelDisposables.push(kernel);
+
+            this.fileInKernel = undefined;
+            await this.runInitialization(kernel, this.owner);
+            this.finishSysInfoMessage(kernel, sysInfoCell, SysInfoReason.Start);
+            kernelPromise.resolve(kernel);
+            return kernel;
+        } catch (ex) {
+            kernelPromise.reject(ex);
+            this.currentKernelInfo.kernel = undefined;
+            this.disconnectKernel();
+            if (this.owner) {
+                // The actual error will be displayed in the cell, hence no need to display the actual
+                // error here, else we'd just be duplicating the error messages.
+                await this.deleteSysInfoCell(sysInfoCell);
+            } else {
+                // We don't have a cell when starting IW without an *.py file,
+                // hence display error where the sysinfo is displayed.
+                await this.finishSysInfoWithFailureMessage(ex, sysInfoCell);
             }
-        }
-    }
-
-    private async kernelEventHook(ev: 'willRestart' | 'willInterrupt') {
```

Comment:
Moved this to an internal fucntion so that its closer to the pre-restart and post re-start event handling.
It was a little weird to have some code inline and rest outside, was difficult to follow what was going on.